File: .temp/flutter_app/lib/domain/use_cases/create_installment_plan_use_case.dart
import 'package:injectable/injectable.dart';
import '../models/installment.dart';
import '../repository/installment_repository.dart';

@lazySingleton
class CreateInstallmentPlanUseCase {
  final InstallmentRepository _installmentRepository;

  CreateInstallmentPlanUseCase(this._installmentRepository);

  Future<void> call({
    required int expenseId,
    required int totalAmount,
    required int durationMonths,
    required int startDate,
  }) async {
    if (durationMonths <= 0) return;

    final monthlyPayment = (totalAmount / durationMonths).floor();

    final installment = Installment(
      expenseId: expenseId,
      totalAmount: totalAmount,
      monthlyPayment: monthlyPayment,
      durationMonths: durationMonths,
      remainingBalance: totalAmount,
      nextDueDate: startDate,
      status: "Active",
      createdAt: DateTime.now().millisecondsSinceEpoch,
    );

    final installmentId = await _installmentRepository.createInstallment(installment);

    await _installmentRepository.createInstallmentItems(
      installmentId: installmentId,
      durationMonths: durationMonths,
      totalAmount: totalAmount,
      startDate: startDate,
    );
  }
}

File: shared/src/commonMain/kotlin/com/sans/finance/domain/usecase/CreateInstallmentPlanUseCase.kt
package com.sans.finance.domain.usecase

import com.sans.finance.domain.model.Installment
import com.sans.finance.domain.repository.InstallmentRepository


class CreateInstallmentPlanUseCase constructor(
    private val installmentRepository: InstallmentRepository
) {
    suspend operator fun invoke(
        expenseId: Long,
        totalAmount: Long,
        durationMonths: Int,
        startDate: Long
    ) {
        if (durationMonths <= 0) return

        val monthlyPayment = totalAmount / durationMonths

        val installment = Installment(
            expenseId = expenseId,
            totalAmount = totalAmount,
            monthlyPayment = monthlyPayment,
            durationMonths = durationMonths,
            remainingBalance = totalAmount,
            nextDueDate = startDate,
            status = "Active"
        )

        val installmentId = installmentRepository.createInstallment(installment)

        installmentRepository.createInstallmentItems(
            installmentId,
            durationMonths,
            totalAmount,
            startDate
        )
    }
}
